readme.md
<p align="center"><a target="_blank" href="https://lsx.lsdev.biz/"><img width="100px;" src="https://lsx.lsdev.biz/wp-content/uploads/2019/03/givewp-payfast-integration.png" alt="LSX PayFast Gateway for Give Extension"></a>
</p>
<h1 align="center">LSX PayFast Gateway for Give</h1>

<p align="center">
	<a href="https://www.gnu.org/licenses/gpl-3.0.en.html"><img src="https://poser.pugx.org/woocommerce/woocommerce/license" alt="license"></a>
	<a href="http://gulpjs.com/"><img src="https://img.shields.io/badge/built%20with-gulp.js-green.svg" alt="Built with gulp.js"></a> 
	<a href="https://travis-ci.org/github/lightspeeddevelopment/lsx-give-payfast-gateway"><img src="https://travis-ci.org/lightspeeddevelopment/lsx-give-payfast-gateway.svg?branch=master" alt="Build Status"></a>
	<a href="https://lsx.lsdev.biz/"><img src="https://lsx.lsdev.biz/wp-content/uploads/2019/06/Designed-for-LSX-Theme-blue.png" alt="Made for LSX"></a>
</p>

## Important notice for existing users

If you previously bought the add-on, the following information is very important. There are a few steps you need to follow when moving to the new version.

* Deactivate and delete the previous plugin completely.
* Then install the new version. Either:Search for the plugin, "LSX PayFast Gateway for Give", via the plugins screen inside your WordPress dashboard
* Or download the plugin from WordPress.org and upload it via your plugins screen

Unfortunately you may need to reconfigure payments via PayFast in Give. The plugin changed in a way that could not avoid this.

## Documentation & Support
Whether you are an existing or new user of this plugin, detailed setup documentation can be found [here](https://www.lsdev.biz/lsx/documentation/lsx-extensions/lsx-payfast-payment-gateway-for-give/)

The plugin is free, but **we do not offer free support**. Please consult our documentation which answers all questions related to setting the plugin up. If it's not working, please follow the steps below. We recommend you test on a development or staging version of your website, most good hosting companies offer a tool that can easily copy your site to a staging version. If not, contact your host directly to set one up.

* Ensure that WordPress, your theme and plugins are fully up to date.
* Activate a default WordPress theme, like twenty sixteen or twenty eighteen
* Disable all plugins but Give and the PayFast Extension
* Test your payments again
* If it works, it means there is a problem with your theme or one of your other plugins that isn't covered by our support policy. Or, there is a problem with your server environment.
* If it still does not work, create a user with administrative access and email that to us at [support@lsdev.biz](mailto:support@lsdev.biz), please include detailed steps on how to reproduce the error, reference any relevant links and include screenshots. We also need to know what operating system and browser you are using.

If you need assistance setting up the plugin, a customization, or it needs to be made compatible with your theme and plugins, please contact us for a quote.

## Contributing to the LSX PayFast Gateway for Give Plugin

If you're a developer who's spotted a bug issue and have a fix, or simply have functionality you think would extend our core theme, we are always happy to accept your contribution! Visit the [LSX PayFast Gateway for Give Plugin on Github](https://github.com/lightspeeddevelopment/lsx-give-payfast-gateway) and submit a Pull Request with your updates.

## Like what you see?
<a href="https://www.lsdev.biz/contact/"><img src="https://www.lsdev.biz/wp-content/uploads/2020/02/work-with-lightspeed.png" width="850" alt="Work with us at LightSpeed"></a>


changelog.md
# Change log

## [[1.2.8]](https://github.com/lightspeeddevelopment/lsx-give-payfast-gateway/releases/tag/1.2.8) - 2023-08-09

### Security
- General testing to ensure compatibility with latest WordPress version (6.3).
- General testing to ensure compatibility with latest Give - Donation Plugin version (2.31.0).
- General testing to ensure compatibility with latest Give - Recurring Donations (1.12.2).

## [[1.2.7]](https://github.com/lightspeeddevelopment/lsx-give-payfast-gateway/releases/tag/1.2.7) - 2023-04-20

### Security
- General testing to ensure compatibility with latest WordPress version (6.2).

## [[1.2.6]](https://github.com/lightspeeddevelopment/lsx-give-payfast-gateway/releases/tag/1.2.6) - 2022-12-23

### Security
- General testing to ensure compatibility with latest WordPress version (6.1.1).
- General testing to ensure compatibility with latest Give - Donation Plugin version (2.23.2).
- General testing to ensure compatibility with latest Give - Recurring Donations (1.12.2).

## [[1.2.5]](https://github.com/lightspeeddevelopment/lsx-give-payfast-gateway/releases/tag/1.2.5) - 2020-08-07

### Fixed
- Update the plugin to use the same signature generating and verification as the [WooCommerce Extension](https://github.com/woocommerce/woocommerce-gateway-payfast/blob/master/includes/class-wc-gateway-payfast.php#L1046)
- Fixed the signature mismatch.

## [[1.2.4]](https://github.com/lightspeeddevelopment/lsx-give-payfast-gateway/releases/tag/1.2.4) - 2020-03-30

### Security
- General testing to ensure compatibility with latest WordPress version (5.4).
- General testing to ensure compatibility with latest LSX Theme version (2.7).

## [[1.2.3]](https://github.com/lightspeeddevelopment/lsx-give-payfast-gateway/releases/tag/1.2.3) - 2019-09-03

### Added
- Adding the .gitattributes file to remove unnecessary files from the WordPress version.


## [[1.2.2]]()

### Fixed
- Fixed not being able to use surnames or names with more than 1 name.


## [[1.2.1]]()

### Changed
- Changed Prefixes.

### Fixed
- Fixed passphrase signature generation


## [[1.2.0]]()

### Deprecated
- Removed Email Sharing Option


## [[1.1.2]]()

### Deprecated
- Removed API Class


## [[1.1.1]]()

### Changed
- Updated the License class.

### Added
- Added in a note for when your license is active.


## [[1.1.0]]()

### Added
- Added recurring donations support.
- Added in PHPDoc Tags for all functions.

### Changed
- Changed the Global $give_options to call the 'give_get_settings' method.

### Deprecated
- Removed the Functions adding the ZAR Currency and Symbol, this is supported Give WP now.
- Removed the PassPhrase from the ITN validation call.


## [[1.0.1]]()

### Added
- Added support for passphrase.

### Fixed
- Corrected the format of the email address when sent to PayFast.


## [[1.0.0]]()

### Added
- Initial Release.


lsx-payfast.php
<?php
/**
 * Plugin Name: Give - PayFast Gateway
 * Plugin URI: https://www.lsdev.biz/product/givewp-payfast-integration-addon/
 * Description: LightSpeed’s PayFast Gateway for GiveWP is the only way to use the powerful Give plugin for WordPress to accept Rands in South Africa. Give is a flexible, robust, and simple WordPress plugin for accepting donations directly on your website.
 * Author: LightSpeed
 * Version: 1.2
 * Author URI: https://www.lsdev.biz/products/
 * License: GPL3+
 * Text Domain: replaceme
 * Domain Path: /languages/

 @package lsx-give-payfast
 **/

/**
 * Includes the PayFast recurring class, if the recurring addon is active
 */
function lsx_give_payfast_recurring() {
	if ( class_exists( 'Give_Recurring' ) ) {
		include_once plugin_dir_path( __FILE__ ) . 'classes/class-lsx-recurring-payfast.php';
	}
}
add_action( 'init', 'lsx_give_payfast_recurring' );

/**
 * PayFast does not need a CC form, so remove it.
 */
add_action( 'give_payfast_cc_form', '__return_false' );

/**
 *    Registers our text domain with WP
 */
function lsx_give_payfast_load_textdomain() {
	load_plugin_textdomain( 'lsx-give-payfast', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );
}
add_action( 'plugins_loaded', 'lsx_give_payfast_load_textdomain' );

/**
 * Registers the gateway
 */
function lsx_give_payfast_register_gateway( $gateways ) {
	$gateways['payfast'] = array(
		'admin_label'    => 'PayFast',
		'checkout_label' => __( 'PayFast', 'lsx-give-payfast' ),
	);
	return $gateways;
}
add_filter( 'give_payment_gateways', 'lsx_give_payfast_register_gateway' );

/**
 * Processes the order and redirect to the PayFast Merchant page
 */
function lsx_give_payfast_process_payment( $purchase_data, $recurring = false ) {
	$lsx_options = give_get_settings();

	// check there is a gateway name.
	if ( ! isset( $purchase_data['post_data']['give-gateway'] ) ) {
		return;
	}

	// collect payment data.
	$payment_data = array(
		'price'           => $purchase_data['price'],
		'give_form_title' => $purchase_data['post_data']['give-form-title'],
		'give_form_id'    => $purchase_data['post_data']['give-form-id'],
		'date'            => $purchase_data['date'],
		'user_email'      => $purchase_data['user_email'],
		'purchase_key'    => $purchase_data['purchase_key'],
		'currency'        => give_get_currency(),
		'user_info'       => $purchase_data['user_info'],
		'status'          => 'pending',
		'gateway'         => 'payfast',
	);
	$required     = array(
		'give_first' => __( 'First Name is not entered.', 'lsx-give-payfast' ),
		'give_last'  => __( 'Last Name is not entered.', 'lsx-give-payfast' ),
	);

	foreach ( $required as $field => $error ) {
		if ( ! $purchase_data['post_data'][ $field ] ) {
			give_set_error( 'billing_error', $error );
		}
	}

	$errors = give_get_errors();

	if ( $errors ) {
		// problems? send back.
		give_send_back_to_checkout( '?payment-mode=' . $purchase_data['post_data']['give-gateway'] );
	} else {

		// not a recurring- do payment insert.
		if ( false === $recurring ) {
			// record the pending payment.
			$payment = give_insert_payment( $payment_data );
			// check payment.
			if ( ! $payment ) {
				// problems? send back.
				give_send_back_to_checkout( '?payment-mode=' . $purchase_data['post_data']['give-gateway'] );
			}
		} else {

			$payment = $recurring->parent_payment_id;
		}

		$total = $purchase_data['price'];

		$seckey = $lsx_options['payfast_customer_id'] . $lsx_options['payfast_key'] . $total;
		$seckey = md5( $seckey );

		if ( give_is_test_mode() ) {
			// test mode.
			$payfast_url = 'https://sandbox.payfast.co.za/eng/process';
		} else {
			// live mode.
			$payfast_url = 'https://www.payfast.co.za/eng/process';
		}

		$redirect     = get_permalink( $lsx_options['success_page'] );
		$query_string = null;

		$permalink = give_get_failed_transaction_uri();
		$cancelurl = add_query_arg( 'error', '', $permalink );

		$payfast_args  = 'merchant_id=' . $lsx_options['payfast_customer_id'];
		$payfast_args .= '&merchant_key=' . $lsx_options['payfast_key'];
		$payfast_args .= '&return_url=' . urlencode( apply_filters( 'lsx_success_page_redirect', $redirect, 'payfast', $query_string ) );
		$payfast_args .= '&cancel_url=' . urlencode( $cancelurl );
		$payfast_args .= '&notify_url=' . urlencode( trailingslashit( home_url() ) );
		$payfast_args .= '&name_first=' . $purchase_data['post_data']['give_first'];
		$payfast_args .= '&name_last=' . $purchase_data['post_data']['give_last'];
		$payfast_args .= '&email_address=' . urlencode( $purchase_data['post_data']['give_email'] );
		$payfast_args .= '&m_payment_id=' . $payment;
		$payfast_args .= '&amount=' . $total;
		$payfast_args .= '&item_name=' . urlencode( $purchase_data['post_data']['give-form-title'] );
		$payfast_args .= '&custom_int1=' . $payment;
		$payfast_args .= '&custom_str1=' . $seckey;

		if ( false !== $recurring ) {
			$payfast_args .= '&custom_str2=' . $recurring->profile_id;
			$payfast_args .= '&subscription_type=1';
			switch ( $purchase_data['period'] ) {
				case 'month':
					$frequency = 3;
					break;
				case 'year':
					$frequency = 6;
					break;
			}
			$payfast_args .= '&frequency=' . $frequency;
			$payfast_args .= '&cycles=' . $purchase_data['times'];

		}

		if ( isset( $lsx_options['payfast_pass_phrase'] ) ) {
			$pass_phrase = trim( $lsx_options['payfast_pass_phrase'] );
		}
		if ( ! empty( $pass_phrase ) ) {
			$payfast_args .= '&pass_phrase=' . urlencode( $pass_phrase );
		}

		update_option( 'first_signature', md5( $payfast_args ) );

		$payfast_args .= '&signature=' . md5( $payfast_args );

		wp_redirect( $payfast_url . '?' . $payfast_args );
		exit();

	}

}
add_action( 'give_gateway_payfast', 'lsx_give_payfast_process_payment' );

/**
 * Processes the order and redirect to the PayFast Merchant page
 */

function lsx_give_payfast_get_realip() {
	$client  = wp_unslash( $_SERVER['HTTP_CLIENT_IP'] );
	$forward = wp_unslash( $_SERVER['HTTP_X_FORWARDED_FOR'] );
	$remote  = wp_unslash( $_SERVER['REMOTE_ADDR'] );

	if ( filter_var( $client, FILTER_VALIDATE_IP ) ) {
		$ip = $client;
	} elseif ( filter_var( $forward, FILTER_VALIDATE_IP ) ) {
		$ip = $forward;
	} else {
		$ip = $remote;
	}

	return $ip;
}

/**
 * An action that handles the call from PayFast to tell Give the order was Completed
 */
function lsx_give_payfast_ipn() {
	$lsx_options = give_get_settings();

	if ( isset( $_REQUEST['m_payment_id'] ) ) {

		if ( give_is_test_mode() ) {
			$pf_host = 'https://sandbox.payfast.co.za/eng/query/validate';
			give_insert_payment_note( $_REQUEST['m_payment_id'], 'ITN callback has been triggered.' );
		} else {
			$pf_host = 'https://www.payfast.co.za/eng/query/validate';
		}

		$pf_error         = false;
		$pf_param_string  = '';
		$validate_string  = '';

		if ( ! $pf_error ) {
			// Strip any slashes in data.
			foreach ( $_POST as $key => $val ) {
				$_POST[ $key ] = stripslashes( $val );
			}
			foreach ( $_POST as $key => $val ) {
				if ( 'signature' != $key ) {
					$pf_param_string .= $key . '=' . urlencode( $val ) . '&';
				}
			}
			$pf_param_string = substr( $pf_param_string, 0, - 1 );
			$validate_string = $pf_param_string;
			if ( isset( $lsx_options['payfast_pass_phrase'] ) ) {
				$pass_phrase = trim( $lsx_options['payfast_pass_phrase'] );
				if ( ! empty( $pass_phrase ) ) {
					$pf_param_string .= '&pass_phrase=' . urlencode( $pass_phrase );
				}
			}
		}
		$signature = md5( $pf_param_string );

		if ( give_is_test_mode() ) {
			// translators:
			give_insert_payment_note( $_REQUEST['m_payment_id'], sprintf( __( 'Signature Returned %1$s. Generated Signature %2$s.', 'lsx-give-payfast' ), $_POST['signature'], $signature ) );
		}

		if ( $signature != $_POST['signature'] ) {
			$pf_error = 'SIGNATURE';
			$error   = array(
				'oursig' => $signature,
				'vars'   => $_POST,
			);
		}

		if ( ! $pf_error ) {
			$valid_hosts = array(
				'www.payfast.co.za',
				'sandbox.payfast.co.za',
				'w1w.payfast.co.za',
				'w2w.payfast.co.za',
			);

			$valid_ips  = array();
			$sender_ip = lsx_give_payfast_get_realip();
			foreach ( $valid_hosts as $pf_hostname ) {
				$ips = gethostbynamel( $pf_hostname );

				if ( false !== $ips ) {
					$valid_ips = array_merge( $valid_ips, $ips );
				}
			}

			$valid_ips = array_unique( $valid_ips );

			if ( ! in_array( $sender_ip, $valid_ips ) ) {
				$pf_error = array(
					'FROM'  => $sender_ip,
					'VALID' => $valid_ips,
				);
			}
		}

		/*
		* If it fails for any reason, add that to the order.
		*/
		if ( false !== $pf_error ) {
			// translators:
			give_insert_payment_note( $_POST['m_payment_id'], sprintf( __( 'Payment Failed. The error is %s.', 'lsx-give-payfast' ), print_r( $pf_error, true ) ) );
		} else {

			$response = wp_remote_post(
				$pf_host, array(
					'method'      => 'POST',
					'timeout'     => 60,
					'redirection' => 5,
					'httpversion' => '1.0',
					'blocking'    => true,
					'headers'     => array(),
					'body'        => $validate_string,
					'cookies'     => array(),
				)
			);

			if ( give_is_test_mode() ) {
				give_insert_payment_note(
					$_POST['m_payment_id'], sprintf(
						// translators:
						__( 'PayFast ITN Params - %1$s %2$s.', 'lsx-give-payfast' ), $pf_host, print_r(
							array(
								'method'      => 'POST',
								'timeout'     => 60,
								'redirection' => 5,
								'httpversion' => '1.0',
								'blocking'    => true,
								'headers'     => array(),
								'body'        => $validate_string,
								'cookies'     => array(),
							), true
						)
					)
				);
				// translators:
				give_insert_payment_note( $_POST['m_payment_id'], sprintf( __( 'PayFast ITN Response. %s.', 'lsx-give-payfast' ), print_r( $response['body'], true ) ) );
			}

			if ( ! is_wp_error( $response ) && ( $response['response']['code'] >= 200 || $response['response']['code'] < 300 ) ) {
				$res = $response['body'];
				if ( false == $res ) {
					$pf_error = $response;

				}
			}
		}

		if ( ! $pf_error ) {
			$lines = explode( "\n", $res );
		}

		if ( ! $pf_error ) {
			$result = trim( $lines[0] );

			if ( strcmp( $result, 'VALID' ) === 0 ) {
				if ( 'COMPLETE' == $_POST['payment_status'] ) {

					if ( ! empty( $_POST['custom_str2'] ) ) {
						$subscription = new Give_Subscription( $_POST['custom_str2'], true );
						// Retrieve pending subscription from database and update it's status to active and set proper profile ID.
						$subscription->update(
							array(
								'profile_id' => $_POST['token'],
								'status'     => 'active',
							)
						);
					}
					give_set_payment_transaction_id( $_POST['m_payment_id'], $_POST['pf_payment_id'] );
					// translators:
					give_insert_payment_note( $_POST['m_payment_id'], sprintf( __( 'PayFast Payment Completed. The Transaction Id is %s.', 'lsx-give-payfast' ), $_POST['pf_payment_id'] ) );
					give_update_payment_status( $_POST['m_payment_id'], 'publish' );

				} else {
					// translators:
					give_insert_payment_note( $_POST['m_payment_id'], sprintf( __( 'PayFast Payment Failed. The Response is %s.', 'lsx-give-payfast' ), print_r( $response['body'], true ) ) );
				}
			}
		}
	}
}
add_action( 'wp_head', 'lsx_give_payfast_ipn' );

/**
 * Registers our PayFast setting with Give.
 *
 * @param  $settings
 * @return array
 */
function lsx_give_payfast_add_settings( $settings ) {

	$payfast_settings = array(

		array(
			'id'   => 'payfast_settings',
			'name' => __( 'PayFast Settings', 'lsx-give-payfast' ),
			'type' => 'give_title',
		),
		array(
			'id'   => 'payfast_api_email',
			'name' => __( 'Email Address', 'lsx-give-payfast' ),
			'desc' => __( 'This is the email address you used to purchase the plugin.', 'lsx-give-payfast' ),
			'type' => 'text',
			'size' => 'regular',
		),
		array(
			'id'   => 'payfast_customer_id',
			'name' => __( 'PayFast Merchant Id', 'lsx-give-payfast' ),
			'desc' => __( 'Please enter your PayFast Merchant Id; this is needed in order to take payment.', 'lsx-give-payfast' ),
			'type' => 'text',
			'size' => 'regular',
		),
		array(
			'id'   => 'payfast_key',
			'name' => __( 'PayFast Key', 'lsx-give-payfast' ),
			'desc' => __( 'Please enter your PayFast Key; this is needed in order to take payment.', 'lsx-give-payfast' ),
			'type' => 'text',
			'size' => 'regular',
		),
		array(
			'id'   => 'payfast_pass_phrase',
			'name' => __( 'pass_phrase', 'lsx-give-payfast' ),
			'desc' => __( 'This is set by yourself in the "Settings" section of the logged in area of the PayFast Dashboard.', 'lsx-give-payfast' ),
			'type' => 'text',
			'size' => 'regular',
		),
	);

	return array_merge( $settings, $payfast_settings );
}
add_filter( 'give_settings_gateways', 'lsx_give_payfast_add_settings' );


give-payfast.php
<?php
/**
 * Plugin Name: LSX PayFast Gateway for Give
 * Plugin URI: https://www.lsdev.biz/product/givewp-payfast-integration-addon/
 * Description: The LSX PayFast Gateway for GiveWP is the only way to use the powerful Give plugin for WordPress to accept Rands in South Africa. Give is a flexible, robust, and simple WordPress plugin for accepting donations directly on your website.
 * Author: LightSpeed
 * Version: 1.2.8
 * Author URI: https://www.lsdev.biz/products/
 * License: GPL3+
 * Text Domain: payfast_give
 * Domain Path: /languages/

 * @package lsx-give-payfast-gateway
 **/

/**
 * Includes the PayFast recurring class, if the recurring addon is active
 */


add_action( 'give_gateway_payfast', 'payfast_process_payment' );

/**
 * Registers the Gateway with the recurring classes.
 *
 * @param  array $gateways
 * @return array
 */
function give_payfast_register_gateway( $gateways ) {
	if ( class_exists( 'Give_Recurring' ) ) {
		include_once plugin_dir_path( __FILE__ ) . 'classes/class-give-recurring-payfast.php';
		$give_recurring_payfast = new Give_Recurring_PayFast();
		$gateways['payfast']    = 'Give_Recurring_PayFast';
	}
	return $gateways;
}
add_action( 'give_recurring_available_gateways', 'give_payfast_register_gateway' );

/**
 * PayFast does not need a CC form, so remove it.
 */
add_action( 'give_payfast_cc_form', '__return_false' );

/**
 *    Registers our text domain with WP
 */
function give_payfast_load_textdomain() {
	load_plugin_textdomain( 'payfast_give', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );
}
add_action( 'plugins_loaded', 'give_payfast_load_textdomain' );

/**
 * Registers the gateway
 */
function payfast_register_gateway( $gateways ) {
	$gateways['payfast'] = array(
		'admin_label'    => 'PayFast',
		'checkout_label' => __( 'PayFast', 'payfast_give' ),
	);
	return $gateways;
}
add_filter( 'give_payment_gateways', 'payfast_register_gateway' );

/**
 * Processes the order and redirect to the PayFast Merchant page
 */
function payfast_process_payment( $purchase_data, $recurring = false ) {
	$give_options = give_get_settings();

	// check there is a gateway name.
	if ( ! isset( $purchase_data['post_data']['give-gateway'] ) ) {
		return;
	}

	// collect payment data.
	$payment_data = array(
		'price'           => $purchase_data['price'],
		'give_form_title' => $purchase_data['post_data']['give-form-title'],
		'give_form_id'    => $purchase_data['post_data']['give-form-id'],
		'date'            => $purchase_data['date'],
		'user_email'      => $purchase_data['user_email'],
		'purchase_key'    => $purchase_data['purchase_key'],
		'currency'        => give_get_currency(),
		'user_info'       => $purchase_data['user_info'],
		'status'          => 'pending',
		'gateway'         => 'payfast',
	);
	$required     = array(
		'give_first' => __( 'First Name is not entered.', 'payfast_give' ),
		'give_last'  => __( 'Last Name is not entered.', 'payfast_give' ),
	);

	foreach ( $required as $field => $error ) {
		if ( ! $purchase_data['post_data'][ $field ] ) {
			give_set_error( 'billing_error', $error );
		}
	}

	$errors = give_get_errors();

	if ( $errors ) {
		// problems? send back.
		give_send_back_to_checkout( '?payment-mode=' . $purchase_data['post_data']['give-gateway'] );
	} else {

		// not a recurring- do payment insert.
		if ( false === $recurring ) {
			// record the pending payment.
			$payment = give_insert_payment( $payment_data );
			// check payment.
			if ( ! $payment ) {
				// problems? send back.
				give_send_back_to_checkout( '?payment-mode=' . $purchase_data['post_data']['give-gateway'] );
			}
		} else {

			$payment = $recurring->parent_payment_id;
		}

		$total = $purchase_data['price'];

		$seckey = $give_options['payfast_customer_id'] . $give_options['payfast_key'] . $total;
		$seckey = md5( $seckey );

		if ( give_is_test_mode() ) {
			// test mode.
			$payfast_url = 'https://sandbox.payfast.co.za/eng/process';
		} else {
			// live mode.
			$payfast_url = 'https://www.payfast.co.za/eng/process';
		}

		$redirect     = get_permalink( $give_options['success_page'] );
		$query_string = null;

		$cancelurl = give_get_failed_transaction_uri();

		if ( give_is_test_mode() ) {
			give_insert_payment_note( $payment, $cancelurl );
		}

		$payfast_args  = 'merchant_id=' . $give_options['payfast_customer_id'];
		$payfast_args .= '&merchant_key=' . $give_options['payfast_key'];
		$payfast_args .= '&return_url=' . urlencode( apply_filters( 'give_success_page_redirect', $redirect, 'payfast', $query_string ) );
		$payfast_args .= '&cancel_url=' . urlencode( $cancelurl );
		$payfast_args .= '&notify_url=' . urlencode( trailingslashit( home_url() ) );
		$payfast_args .= '&name_first=' . urlencode( $purchase_data['post_data']['give_first'] );
		$payfast_args .= '&name_last=' . urlencode( $purchase_data['post_data']['give_last'] );
		$payfast_args .= '&email_address=' . urlencode( $purchase_data['post_data']['give_email'] );
		$payfast_args .= '&m_payment_id=' . $payment;
		$payfast_args .= '&amount=' . $total;
		$payfast_args .= '&item_name=' . urlencode( $purchase_data['post_data']['give-form-title'] );
		$payfast_args .= '&custom_int1=' . $payment;
		$payfast_args .= '&custom_str1=' . $seckey;

		if ( false !== $recurring ) {
			$payfast_args .= '&custom_str2=' . $recurring->profile_id;
			$payfast_args .= '&subscription_type=1';
			switch ( $purchase_data['period'] ) {
				case 'month':
					$frequency = 3;
					break;
				case 'year':
					$frequency = 6;
					break;
			}
			$payfast_args .= '&frequency=' . $frequency;
			$payfast_args .= '&cycles=' . $purchase_data['times'];

		}

		if ( isset( $give_options['payfast_pass_phrase'] ) ) {
			$pass_phrase = trim( $give_options['payfast_pass_phrase'] );
		}
		$signature_str = $payfast_args;
		if ( ! empty( $pass_phrase ) ) {
			$signature_str .= '&passphrase=' . urlencode( $pass_phrase );
		}

		$payfast_args .= '&signature=' . md5( $signature_str );

		if ( give_is_test_mode() && function_exists( 'give_record_log' ) ) {
			give_record_log( 'Payfast - #' . $payment, $payfast_args, 0, 'api_requests' );
			give_insert_payment_note( $payment, $payfast_args );
		}

		wp_redirect( $payfast_url . '?' . $payfast_args );
		exit();

	}
}

/**
 * Processes the order and redirect to the PayFast Merchant page
 */

function payfast_get_realip() {
	$client  = wp_unslash( $_SERVER['HTTP_CLIENT_IP'] );
	$forward = wp_unslash( $_SERVER['HTTP_X_FORWARDED_FOR'] );
	$remote  = wp_unslash( $_SERVER['REMOTE_ADDR'] );

	if ( filter_var( $client, FILTER_VALIDATE_IP ) ) {
		$ip = $client;
	} elseif ( filter_var( $forward, FILTER_VALIDATE_IP ) ) {
		$ip = $forward;
	} else {
		$ip = $remote;
	}

	return $ip;
}

/**
 * An action that handles the call from PayFast to tell Give the order was Completed
 */
function payfast_ipn() {
	if ( function_exists( 'give_get_settings' ) ) {
		$give_options = give_get_settings();

		if ( isset( $_REQUEST['m_payment_id'] ) ) {

			if ( give_is_test_mode() ) {
				$pf_host = 'https://sandbox.payfast.co.za/eng/query/validate';
				give_insert_payment_note( $_REQUEST['m_payment_id'], 'ITN callback has been triggered.' );
			} else {
				$pf_host = 'https://www.payfast.co.za/eng/query/validate';
			}

			$pf_error         = false;
			$pf_param_string  = '';
			$validate_string  = '';

			if ( ! $pf_error ) {
				// Strip any slashes in data.
				foreach ( $_POST as $key => $val ) {
					$_POST[ $key ] = stripslashes( $val );
				}
				foreach ( $_POST as $key => $val ) {
					if ( 'signature' != $key ) {
						$pf_param_string .= $key . '=' . urlencode( $val ) . '&';
					}
				}
				$pf_param_string = substr( $pf_param_string, 0, - 1 );
				$validate_string = $pf_param_string;
				if ( isset( $give_options['payfast_pass_phrase'] ) ) {
					$pass_phrase = trim( $give_options['payfast_pass_phrase'] );
					if ( ! empty( $pass_phrase ) ) {
						$pf_param_string .= '&passphrase=' . urlencode( $pass_phrase );
					}
				}
			}
			$signature = md5( $pf_param_string );

			if ( give_is_test_mode() ) {
				// translators:
				give_insert_payment_note( $_REQUEST['m_payment_id'], sprintf( __( 'Signature Returned %1$s. Generated Signature %2$s.', 'payfast_give' ), $_POST['signature'], $signature ) );
			}

			if ( $signature != $_POST['signature'] ) {
				$pf_error = 'SIGNATURE';
				$error   = array(
					'oursig' => $signature,
					'vars'   => $_POST,
				);
			}

			if ( ! $pf_error ) {
				$valid_hosts = array(
					'www.payfast.co.za',
					'sandbox.payfast.co.za',
					'w1w.payfast.co.za',
					'w2w.payfast.co.za',
				);

				$valid_ips  = array();
				$sender_ip = payfast_get_realip();
				foreach ( $valid_hosts as $pf_hostname ) {
					$ips = gethostbynamel( $pf_hostname );

					if ( false !== $ips ) {
						$valid_ips = array_merge( $valid_ips, $ips );
					}
				}

				$valid_ips = array_unique( $valid_ips );

				if ( ! in_array( $sender_ip, $valid_ips ) ) {
					$pf_error = array(
						'FROM'  => $sender_ip,
						'VALID' => $valid_ips,
					);
				}
			}

			/*
			* If it fails for any reason, add that to the order.
			*/
			if ( false !== $pf_error ) {
				// translators:
				give_insert_payment_note( $_POST['m_payment_id'], sprintf( __( 'Payment Failed. The error is %s.', 'payfast_give' ), print_r( $pf_error, true ) ) );
			} else {

				$response = wp_remote_post(
					$pf_host, array(
						'method'      => 'POST',
						'timeout'     => 60,
						'redirection' => 5,
						'httpversion' => '1.0',
						'blocking'    => true,
						'headers'     => array(),
						'body'        => $validate_string,
						'cookies'     => array(),
					)
				);

				if ( give_is_test_mode() ) {
					give_insert_payment_note(
						$_POST['m_payment_id'], sprintf(
							// translators:
							__( 'PayFast ITN Params - %1$s %2$s.', 'payfast_give' ), $pf_host, print_r(
								array(
									'method'      => 'POST',
									'timeout'     => 60,
									'redirection' => 5,
									'httpversion' => '1.0',
									'blocking'    => true,
									'headers'     => array(),
									'body'        => $validate_string,
									'cookies'     => array(),
								), true
							)
						)
					);
					// translators:
					give_insert_payment_note( $_POST['m_payment_id'], sprintf( __( 'PayFast ITN Response. %s.', 'payfast_give' ), print_r( $response['body'], true ) ) );
				}

				if ( ! is_wp_error( $response ) && ( $response['response']['code'] >= 200 || $response['response']['code'] < 300 ) ) {
					$res = $response['body'];
					if ( false == $res ) {
						$pf_error = $response;

					}
				}
			}

			if ( ! $pf_error ) {
				$lines = explode( "\n", $res );
			}

			if ( ! $pf_error ) {
				$result = trim( $lines[0] );

				if ( strcmp( $result, 'VALID' ) === 0 ) {
					if ( 'COMPLETE' == $_POST['payment_status'] ) {

						if ( ! empty( $_POST['custom_str2'] ) ) {
							$subscription = new Give_Subscription( $_POST['custom_str2'], true );
							// Retrieve pending subscription from database and update it's status to active and set proper profile ID.
							$subscription->update(
								array(
									'profile_id' => $_POST['token'],
									'status'     => 'active',
								)
							);
						}
						give_set_payment_transaction_id( $_POST['m_payment_id'], $_POST['pf_payment_id'] );
						// translators:
						give_insert_payment_note( $_POST['m_payment_id'], sprintf( __( 'PayFast Payment Completed. The Transaction Id is %s.', 'payfast_give' ), $_POST['pf_payment_id'] ) );
						give_update_payment_status( $_POST['m_payment_id'], 'publish' );

					} else {
						// translators:
						give_insert_payment_note( $_POST['m_payment_id'], sprintf( __( 'PayFast Payment Failed. The Response is %s.', 'payfast_give' ), print_r( $response['body'], true ) ) );
					}
				}
			}
		}
	}

}
add_action( 'wp_head', 'payfast_ipn' );

/**
 * Registers our PayFast setting with Give.
 *
 * @param  $settings
 * @return array
 */
function payfast_add_settings( $settings ) {

	$payfast_settings = array(

		array(
			'id'   => 'payfast_settings',
			'name' => __( 'PayFast Settings', 'payfast_give' ),
			'type' => 'give_title',
		),
		array(
			'id'   => 'payfast_customer_id',
			'name' => __( 'PayFast Merchant ID', 'payfast_give' ),
			'desc' => __( 'Please enter your PayFast Merchant Id; this is needed in order to take payment.', 'payfast_give' ),
			'type' => 'text',
			'size' => 'regular',
		),
		array(
			'id'   => 'payfast_key',
			'name' => __( 'PayFast Key', 'payfast_give' ),
			'desc' => __( 'Please enter your PayFast Key; this is needed in order to take payment.', 'payfast_give' ),
			'type' => 'text',
			'size' => 'regular',
		),
		array(
			'id'   => 'payfast_pass_phrase',
			'name' => __( 'Account Passphrase', 'payfast_give' ),
			'desc' => __( 'This is set by yourself in the "Settings" section of the logged in area of the PayFast Dashboard.', 'payfast_give' ),
			'type' => 'text',
			'size' => 'regular',
		),
	);

	return array_merge( $settings, $payfast_settings );
}
add_filter( 'give_settings_gateways', 'payfast_add_settings' );
